File: scripts/pdf-parser/parse.py
# /// script
# requires-python = ">=3.9"
# dependencies = [
#     "pymupdf",
# ]
# ///

import fitz
import re
import json
import os
import sys

# Windows console encoding fix
sys.stdout.reconfigure(encoding='utf-8')

def run():
    pdf_path = "../../機車駕照筆試題庫(全部804題)-1150218.pdf"
    doc = fitz.open(pdf_path)
    
    categories = [
        {"id": "cat_0", "name": "正確觀念與態度", "icon": "⚖️", "color": "#4f46e5", "bg": "#eef2ff", "desc": "正確觀念與態度"},
        {"id": "cat_1", "name": "主動停讓文化", "icon": "🛡️", "color": "#0891b2", "bg": "#ecfeff", "desc": "主動停讓文化"},
        {"id": "cat_2", "name": "安全駕駛能力", "icon": "🚨", "color": "#16a34a", "bg": "#f0fdf4", "desc": "安全駕駛能力"},
    ]
    current_category = categories[0]
    
    questions = []
    next_expected_id = 1
    
    img_out_dir = "../../public/images"
    os.makedirs(img_out_dir, exist_ok=True)
    
    for page_num in range(len(doc)):
        page = doc.load_page(page_num)
        blocks = page.get_text("dict")["blocks"]
        
        # 1. Collect question text blocks on this page
        page_qs = []
        if questions:
            page_qs.append(questions[-1])
        for b in blocks:
            if b["type"] == 0:
                text = ""
                for l in b["lines"]:
                    text += "".join(s["text"] for s in l["spans"]) + "\n"
                text = text.strip()
                lines = text.split('\n')
                for line in lines:
                    line = line.strip()
                    if not line:
                        continue
                    
                    # Skip headers and footers
                    if '機車駕照筆試題庫' in line or re.match(r'^—\s*\d+\s*—$', line) or '題號 答案' in line or '題目內容' in line or '【 題庫索引 】' in line or line == '分類':
                        continue
                    if re.match(r'^━+$', line):
                        continue
                        
                    # Category detection
                    for c in categories:
                        if c["name"] in line and len(line) < 20:
                            current_category = c
                            break
                            
                    q_match = re.match(r'^(\d+)\s+([123])\s*(.*)$', line)
                    if q_match and int(q_match.group(1)) == next_expected_id:
                        q_id = next_expected_id
                        ans = int(q_match.group(2)) - 1
                        raw_text = q_match.group(3) + " "
                        
                        q_obj = {
                            "id": q_id,
                            "answer": ans,
                            "categoryId": current_category["id"],
                            "rawText": raw_text,
                            "image": None,
                            "y0": b["bbox"][1] # Store Y coordinate
                        }
                        page_qs.append(q_obj)
                        questions.append(q_obj)
                        next_expected_id += 1
                    elif line == str(next_expected_id):
                        # In case ID is on its own line
                        q_obj = {
                            "id": next_expected_id,
                            "answer": None,
                            "categoryId": current_category["id"],
                            "rawText": "",
                            "image": None,
                            "y0": b["bbox"][1],
                            "awaiting_answer": True
                        }
                        page_qs.append(q_obj)
                        questions.append(q_obj)
                        next_expected_id += 1
                    else:
                        # Append text to the last question globally
                        if questions:
                            last_q = questions[-1]
                            if last_q.get("awaiting_answer"):
                                ans_match = re.match(r'^([123])\s*(.*)$', line)
                                if ans_match:
                                    last_q["answer"] = int(ans_match.group(1)) - 1
                                    last_q["rawText"] += ans_match.group(2) + " "
                                    last_q["awaiting_answer"] = False
                                elif re.match(r'^[123]$', line):
                                    last_q["answer"] = int(line) - 1
                                    last_q["awaiting_answer"] = False
                            else:
                                last_q["rawText"] += line + " "

        # 2. Collect images on this page and map to closest question by Y coord
        images = page.get_images(full=True)
        for img in images:
            xref = img[0]
            name = img[7]
            try:
                bbox = page.get_image_bbox(name)
                img_y0 = bbox[1]
                
                # Find closest question
                closest_q = None
                min_dist = float('inf')
                for q in page_qs:
                    dist = abs(q["y0"] - img_y0)
                    if dist < min_dist:
                        min_dist = dist
                        closest_q = q
                        
                if closest_q and min_dist < 100: # Ensure it's reasonably close
                    # Extract image bytes
                    base_image = doc.extract_image(xref)
                    image_bytes = base_image["image"]
                    ext = base_image["ext"]
                    
                    img_filename = f"q_{closest_q['id']}.{ext}"
                    img_path = os.path.join(img_out_dir, img_filename)
                    with open(img_path, "wb") as f:
                        f.write(image_bytes)
                        
                    closest_q["image"] = f"/images/{img_filename}"
            except Exception as e:
                print(f"Failed to process image {name}: {e}")

    # Process options
    final_questions = []
    for q in questions:
        raw = q["rawText"]
        opt_match = re.search(r'(.*?)[\(（]1[\)）]\s*(.*?)[\(（]2[\)）]\s*(.*?)[\(（]3[\)）]\s*(.*)', raw)
        if opt_match:
            text = opt_match.group(1).strip()
            o1 = re.sub(r'[。\.]$', '', opt_match.group(2).strip())
            o2 = re.sub(r'[。\.]$', '', opt_match.group(3).strip())
            o3 = re.sub(r'[。\.]$', '', opt_match.group(4).strip())
            
            final_questions.append({
                "id": q["id"],
                "category": q["categoryId"],
                "text": text,
                "options": [o1, o2, o3],
                "answer": q["answer"] if q["answer"] is not None else 0,
                "image": q["image"],
                "isCore": (q["id"] % 10 == 0),
                "explanation": ""
            })
        else:
            print(f"Failed to parse options for Q: {q['id']} - {raw}")

    js_content = f"""// ============================================================
//  台灣機車駕照筆試完整題庫 — {len(final_questions)}題 (含圖片)
// ============================================================

export const CATEGORIES = {json.dumps(categories, ensure_ascii=False, indent=2)};

export const QUESTIONS = {json.dumps(final_questions, ensure_ascii=False, indent=2)};
"""
    out_path = "../../src/data/questions.js"
    with open(out_path, "w", encoding="utf-8") as f:
        f.write(js_content)
        
    img_count = sum(1 for q in final_questions if q['image'])
    print(f"Parsed {len(final_questions)} questions with {img_count} images.")

if __name__ == "__main__":
    run()

File: scripts/pdf-parser/parse.js
import fs from 'fs'
import path from 'path'
import pdf from 'pdf-parse/lib/pdf-parse.js'
import { fileURLToPath } from 'url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

const pdfPath = path.resolve(__dirname, '../../機車駕照筆試題庫(全部804題)-1150218.pdf')
const outPath = path.resolve(__dirname, '../../src/data/questions.js')

const COLORS = [
  { color: '#4f46e5', bg: '#eef2ff' }, // indigo
  { color: '#0891b2', bg: '#ecfeff' }, // cyan
  { color: '#16a34a', bg: '#f0fdf4' }, // green
  { color: '#dc2626', bg: '#fef2f2' }, // red
  { color: '#d97706', bg: '#fffbeb' }, // amber
  { color: '#8b5cf6', bg: '#f5f3ff' }, // violet
  { color: '#ec4899', bg: '#fdf2f8' }  // pink
]

const ICONS = ['⚖️', '🛡️', '🚨', '🏍️', '🚦', '🛑', '⚠️']

async function run() {
  const dataBuffer = fs.readFileSync(pdfPath)
  const data = await pdf(dataBuffer)
  
  const text = data.text
  const lines = text.split('\n').map(l => l.trim()).filter(l => l.length > 0)
  
  const categories = []
  const questions = []
  
  let currentCategory = null
  let currentQuestion = null
  let nextExpectedId = 1
  
  // Regex to detect category header, e.g. ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n正確觀念與態度\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  // But wait, after pdf-parse, it might just look like text. We can detect categories if we see lines without numbers and not options.
  // Actually, let's just use the exact strings from the index if possible.
  // Let's print out the text first to debug if needed, but we can do a robust parser.
  
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]
    
    // Ignore headers/footers
    if (line.includes('機車駕照筆試題庫') || line.match(/^—\s*\d+\s*—$/) || line.includes('題號 答案') || line.includes('題目內容') || line.includes('【 題庫索引 】') || line.includes('分類')) {
      continue
    }

    // Is it a new question start?
    let qMatch = line.match(/^(\d+)\s+([123])\s*(.*)$/)
    if (qMatch && parseInt(qMatch[1]) === nextExpectedId) {
      if (currentQuestion) {
        processQuestion(currentQuestion, questions)
      }
      currentQuestion = {
        id: nextExpectedId,
        answer: parseInt(qMatch[2]) - 1, // 0-indexed
        categoryId: currentCategory ? currentCategory.id : 'default',
        rawText: qMatch[3] + ' '
      }
      nextExpectedId++
      continue
    }

    if (line === nextExpectedId.toString()) {
      if (currentQuestion) {
        processQuestion(currentQuestion, questions)
      }
      
      const nextLine = lines[i+1]
      let ans, rawText, skip
      
      if (nextLine.match(/^[123]$/)) {
        ans = parseInt(nextLine)
        rawText = ''
        skip = 1
      } else {
        const m = nextLine.match(/^([123])\s+(.*)/)
        if (m) {
          ans = parseInt(m[1])
          rawText = m[2] + ' '
          skip = 1
        } else {
          // not matched, might not be a question start
          continue
        }
      }
      
      currentQuestion = {
        id: nextExpectedId,
        answer: ans - 1, // 0-indexed
        categoryId: currentCategory ? currentCategory.id : 'default',
        rawText: rawText
      }
      nextExpectedId++
      i += skip // Skip the answer line
      continue
    }
    
    // Is it a category header? "正確觀念與態度" surrounded by "━━━..."
    if (line.match(/^━+$/)) {
      // Look at next line
      if (i + 1 < lines.length && !lines[i+1].match(/^━+$/)) {
        const catName = lines[i+1].trim()
        if (catName.length > 0 && !catName.match(/^\d/)) {
          // It's a category!
          let catId = 'cat_' + categories.length
          const catColor = COLORS[categories.length % COLORS.length]
          const catIcon = ICONS[categories.length % ICONS.length]
          
          categories.push({
            id: catId,
            name: catName,
            icon: catIcon,
            color: catColor.color,
            bg: catColor.bg,
            desc: catName
          })
          currentCategory = categories[categories.length - 1]
          i++ // skip the name line
          // skip the next line if it's also '━━━...'
          if (i + 1 < lines.length && lines[i+1].match(/^━+$/)) {
             i++
          }
        }
      }
      continue
    }
    
    if (currentQuestion) {
      currentQuestion.rawText += line + ' '
    }
  }
  
  if (currentQuestion) {
    processQuestion(currentQuestion, questions)
  }
  
  // Clean up categories: some categories might be sub-categories. 
  // If we couldn't parse categories perfectly, fallback to one default.
  if (categories.length === 0) {
    categories.push({ id: 'default', name: '全庫綜合', icon: '🏍️', color: '#4f46e5', bg: '#eef2ff', desc: '全部考題' })
  }

  // Deduplicate categories if needed
  const uniqueCats = []
  const catMap = new Map()
  for (const c of categories) {
    if (!catMap.has(c.name)) {
      catMap.set(c.name, c.id)
      uniqueCats.push(c)
    }
  }
  
  // Fix question category IDs
  for (const q of questions) {
    const c = categories.find(cat => cat.id === q.category)
    if (c && catMap.has(c.name)) {
      q.category = catMap.get(c.name)
    }
  }

  const jsContent = `// ============================================================
//  台灣機車駕照筆試完整題庫 — ${questions.length}題
// ============================================================

export const CATEGORIES = ${JSON.stringify(uniqueCats, null, 2)}

export const QUESTIONS = ${JSON.stringify(questions, null, 2)}
`
  
  fs.writeFileSync(outPath, jsContent, 'utf-8')
  console.log('Parsed ' + questions.length + ' questions and ' + uniqueCats.length + ' categories.')
}

function processQuestion(q, questionsList) {
  // Extract options (1)... (2)... (3)...
  const optMatch = q.rawText.match(/(.*?)[\(（]1[\)）]\s*(.*?)[\(（]2[\)）]\s*(.*?)[\(（]3[\)）]\s*(.*)/)
  if (optMatch) {
    let text = optMatch[1].trim()
    
    // Skip image questions (which have no text in the PDF parsed output)
    if (!text) return;
    
    let o1 = optMatch[2].trim()
    let o2 = optMatch[3].trim()
    let o3 = optMatch[4].trim()
    
    // Remove trailing periods from options
    o1 = o1.replace(/[。\.]$/, '')
    o2 = o2.replace(/[。\.]$/, '')
    o3 = o3.replace(/[。\.]$/, '')
    
    questionsList.push({
      id: q.id,
      category: q.categoryId,
      text: text,
      options: [o1, o2, o3],
      answer: q.answer,
      isCore: (q.id % 10 === 0),
      explanation: ''
    })
  } else {
    // If we fail to parse options, push a dummy and log it
    console.error('Failed to parse options for Q: ' + q.id, q.rawText)
  }
}

run().catch(console.error)
